skxray/fitting/xrf_model.py

```diff
@@ -0,0 +1,1059 @@
+
+    Parameters
+    ----------
+    param_name : str
+        parameter used for fitting
+    input_dict : dict
+        all the initial values and constraints for given parameters
+    input_model : object
+        model object used in lmfit
+    log_option : bool
+        option for logger
+    """
+    if input_dict['bound_type'] == 'none':
+        input_model.set_param_hint(name=param_name, value=input_dict['value'], vary=True)
+    elif input_dict['bound_type'] == 'fixed':
+        input_model.set_param_hint(name=param_name, value=input_dict['value'], vary=False)
+    elif input_dict['bound_type'] == 'lohi':
+        input_model.set_param_hint(name=param_name, value=input_dict['value'], vary=True,
+                                   min=input_dict['min'], max=input_dict['max'])
+    elif input_dict['bound_type'] == 'lo':
+        input_model.set_param_hint(name=param_name, value=input_dict['value'], vary=True,
+                                   min=input_dict['min'])
+    elif input_dict['bound_type'] == 'hi':
+        input_model.set_param_hint(name=param_name, value=input_dict['value'], vary=True,
+                                   min=input_dict['max'])
+    else:
+        raise ValueError("could not set values for {0}".format(param_name))
+    if log_option:
+        logger.debug(' {0} bound type: {1}, value: {2}, range: {3}'.
+                    format(param_name, input_dict['bound_type'], input_dict['value'],
+                           [input_dict['min'], input_dict['max']]))
+
+
+def update_parameter_dict(xrf_parameter, fit_results):
+    """
+    Update fitting parameters dictionary according to given fitting results,
+    usually obtained from previous run.
+    .. warning :: This function mutates the input values.
+
+    Parameters
+    ----------
+    xrf_parameter : dict
+        saving all the fitting values and their bounds
+    fit_results : object
+        ModelFit object from lmfit
+    """
+    for k, v in six.iteritems(xrf_parameter):
+        if k in fit_results.values:
+            xrf_parameter[str(k)]['value'] = fit_results.values[str(k)]
+
+
+_STRATEGY_REGISTRY = {'linear': sfb_pd.linear,
+                      'adjust_element': sfb_pd.adjust_element,
+                      'e_calibration': sfb_pd.e_calibration,
+                      'fit_with_tail': sfb_pd.fit_with_tail,
+                      'free_more': sfb_pd.free_more}
+
+
+def register_strategy(key, strategy, overwrite=True):
+    """
+    Register new strategy.
+
+    Parameters
+    ----------
+    key : str
+        strategy name
+    strategy : dict
+        bound for every parameter
+    """
+    if (not overwrite) and (key in _STRATEGY_REGISTRY):
+            if _STRATEGY_REGISTRY[key] is strategy:
+                return
+            raise RuntimeError("You are trying to overwrite an "
+                               "existing strategy: {}".format(key))
+    _STRATEGY_REGISTRY[key] = strategy
+
```

Comment:
And I stole the idea from https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/legend.py#L498 and there about.